joseflegal/dev-test.joseflegal.com
# Nourena::Josef-Take-Home-Challenge

### Frontend:

1. Created a FileList component which takes an array of files as props
2. Added a computed property that takes the files and filters them by a tag field selected from a drop down then sorts it by date.
3. Added slots to FileList component for the file's image, description and date
4. formatted the date displayed
5. Created FileItem component which call the slots from FileList and display each file's data in a card
6. Used `src/styles/resources/_variables.scss` and `src/styles/resources/_colours.scss` to style FileList and FileItem components and Layout page

![Screenshot (8)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/ea293f9e-e6ed-4f4c-9bd2-205fb84d3fa3)
![Screenshot (9)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/9f82955f-756b-463e-8b4a-c636acd8026a)

### Backend:

-Finished the checkGroup function in the src/views/Rules.vue file, so it returns true if all rules and groups apply, otherwise false.
![Screenshot (10)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/d57b4396-7684-463a-891f-88362564659a)

**Extra credit**:

1. Migrated the data that was managed locally in Rules.vue and Layout.vue to vuex store
2. Write unit tests for checkGroup and checkRule methods in Rules.vue file and for foramtDate in FileList component

### QA:

-  API Testing: 

![Screenshot (1)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/47b2024a-4653-4bb2-8dad-b4c8a340c7a3)
![Screenshot (2)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/f59f1788-7fb2-4496-ad86-08fc794b9640)
![Screenshot (3)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/ec2e6150-9a2b-4d00-9456-1a381c6e1925)
![Screenshot (4)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/4d5bfc94-749e-4445-91ed-cca3e0cd93e9)
![Screenshot (5)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/a9aa3454-98ce-41f7-a6e0-456c11ae7cff)
![Screenshot (6)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/06c2e8c6-9bb4-4902-accf-2499f45928aa)

- Cypress: (E2E tests):

- Please run `npm run cypress` to see E2E tests in cypress:
![Screenshot (11)](https://github.com/joseflegal/dev-test.joseflegal.com/assets/65375174/c87a00e1-ac0c-4f0c-a3b6-0d2d26f50d5d)

- Lighthouse:

[Lighthouse_report.pdf](https://github.com/joseflegal/dev-test.joseflegal.com/files/14215289/Lighthouse_report.pdf)

-In addition, we can add integration with security testing tools such as SonarQube can help identify security vulnerabilities early in the development lifecycle. Security testing tools can scan code for common security issues and enforce secure coding practices.
